### DISCUSSION TITLE
Crank/Buzz settings

### DISCUSSION THREAD

**Discussion by Gubbledenut** [bazmonk/digigurdy-baz#66] [Q&A]
Rather than confuse the face book page I though tI'd  bring the settings I've been playing around with over here.
Using a 4.1 teensy

**SPIN_SAMPLES = 800000;** 

 ( I did try 40000, 60000, 100000, 120000, this value with a rectifier seems to give lowest not jitter cranking)

**const int BUZZ_SMOOTHING = 25; //tried 15....No control
const int BUZZ_DECAY = 10; //was 1...5 seems to work well with 25 above**

( with decay at one coup control was impossible, at 5 it was getting there, but 10 seems to cut off and be Reilly controllable, dropping the buzz smoothing doesn't appear to help with coup much, I will test more with different ratios. I tend to currently think a quite close ratio between the two works best.....work in progress...not that my playing is that good anyway.

**Comment by bazmonk** [comment]
Smart, easier to talk about it here.

Behind the scenes here I haven't stopped fiddling with the circuit.  I feel like I'm missing something electronic-y here.

Right now I've got a rectifier made of possibly-fake 1n34a diodes supposedly meant for use in crystal radios.  With just the rectifier and the resistor over the crank terminals for physical resistance, a multimeter shows the rectifier is very sensitive.  If I put a 100Ohm resistor inline (to keep the current low) it's still as sensitive.  But if I hook it up to the teensy, it wigs out and give me values of like 900 (2-ish volts) constantly.

If I add in a voltage divider, I get crappy less-sensitive response that we've been grappling with.  

The voltage is not too high without the divider, but I can't seem to figure out how to wire the damn thing up without one.  It seems like it "needs" to have resistance there for some reason that's eluding me.  Something behaves differently under load, maybe?  Short of lots of trial and error I'm not sure where I'm going wrong here.

To that end, though, I'm going to dismantle things enough to run it all through a breadboard and test in-situ.  

---

I think the bottom line with the reversible crank is that with the ones we've come up with so far, the crank just doesn't register the low-voltage slow cranking that we need it to.  I fear wrestling with the variables isn't going to achieve it.  A smooth cranking motion will be detrimental couping, and good coups requires your normal cranking to be a little too fast.  It's a pickle, a sort of signal-to-noise ratio problem.

---

The sample/spin/buzz settings... 800000 is very long.  That determines how "long" all the other settings are as it makes each cycle longer.  There's a diminishing return to setting it higher beyond a point:

* Behind the scenes here, the ADC is sampling the analog pin, and it does not sample as fast as we look at the value.  If you actually print out every continuousRead() it does, you'll see that it'll report several of the same value before giving a new one and reporting that several times.  So beyond a point, we are just _checking the same voltage over and over again_.
* I intended the spin_samples to be used to change the timing.  By needlessly checking the voltage more, we're effectively making each cycle take the same amount of time.  I don't know if the Teensy4.1 ADC can poll faster or not... what I think I'm adjusting for is the faster CPU making the rest of the code faster, and slowing down the polling to compensate.  The alternative to doing this is an explicit delay in the code that compensates (still an option, I'm not married to the technique we're doing now).
* The longer cycle accentuates how sensitive all the spin and buzz settings are.  You'd need to reduce them to get the same kind of sensitivity... but then you're back to getting the jitters.

I wonder if a rolling-average (this cycles crank voltage is the average of the new reading and previous ones) would do a better job of smoothing, but I suspect it wouldn't be enough to hide the funky crank sensitivity.

I'm still holding on to hope of coming up with a physical design that works as well as they do w/o the rectifier.  The solution, for me at least, needs to be something that works with minimal adjustment compared to the cranks most digigurdy owners have.

**Comment by bazmonk** [comment]
I think I might be on to something... 7ohms resistance between the crank terminals, in34a diodes, 100ohms coming off the positive side, and 6.8k ohms making up the other half of the voltage divider.  On the breadboard it felt better.

**Reply by Gubbledenut** [reply]
Is the voltage divider going to make any difference to performance or detection? 
Are you suggesting we need to aim for a more precise voltage range when cranking the motor? 
I’ll test the relative number setups in the code agsin…and reduce the 80000, perhaps lower doesn’t make much difference…..ah fun times

**Reply by bazmonk** [reply]
I'm not sure... here's what I noticed:

I had set up a 100Ohm resistor in-line, without a second to make a voltage divider. Like I said, the multimeter looked fantastic, very sensitive, but hooked up to the teensy the voltage would "wig out" and peg itself at like 932 or something.  I think there needs to be some sort of resistance there, for reasons I don't comprehend.

While I was too lazy to solder up another resistor to put the divider back, I just manually twisted on resistors to try things out.  Using 100Ohms as the second resistor (splitting the voltage in half) gave poor sensitivity.  Higher resistors here (less voltage being divided out) seemed to help, so then I tried a super-high one (the highest I had, like 100K Ohms), and it behaved as if it wasn't even there (I'm guessing the resistance was just too high to matter with this current).

I figured if the higher resistors/no resistor made it seem sensitive on the multimeter, and low ones were not sensitive enough, I should be using the highest resistor here that I can that still acts like a resistor is there.  That ended up being 6.8k Ohms out of the variety I have here (didn't try chaining together smaller ones to get more options, just stuck with this).

That's currently what I'm using, and I'm not gonna change it again because I have convinced myself that this is as good as this motor is going to be with this approach.

---

I also cut out a *lot* of the noise in the readings by grounding the casing of the motor to the - / GND side of the circuit (just after the voltage divider and 6.8k resistor).  I'm trying to keep the voltage threshold as absolutely low as possible to get as much response as I can, and I noticed that a value of 3 or 4 would randomly create phantom cranks because of the noise.  Then I realized the noise goes way up when I touch the metal of the crank arm.  Then I realized that if I also touch the metal of the USB cable, or my laptop (which has been powering it as I'm testing), it would go away almost entirely... it's interference, like buzz on a guitar pickup!  It also would go away if I stopped charging my laptop at the same time.  I think it's literally AC buzz that makes up most of the noise.  So instead of me touching my laptop and crank at the same time, I soldered a wire to the metal of the motor and the ground wire going out.  That allows me to lower the threshold even more... I'm using  `V_THRESHOLD = 2`!!!

Highly recommended.

---

So I posted v1.3.0 in the testing branch.  I went in a different direction here...

* There's a `SPIN_STOP_THRESHOLD` now: this is the point in the spin count that it will cut off a note.  So now there is a middle range in the spin where notes that are on stay on, and notes that are off stay off.
* I went with much lower spin samples (1000 on teensy3.5, about 6.3kHz cycles or 150ms per 1,000 loops), and raised up the spin weight/total/decay/etc. values instead.  This was to get more control over the spin algorithm: basically so I could fine-tune the `SPIN_DECAY` value.  Higher spin samples just blindly average more and more readings, and I don't think it's helping much in terms of crank performance.
* The `crank_voltage` being used is now a rolling average with the previous value to smooth it out even more.  Rather than a larger spin sample and an average of that, this smooths out each reading with the one before it.  With this, and the grounded motor, my real voltage sits at 0 and hops up to 1-2, and the smoothed value sits at 0 or 1 (with my v-threshold set at 2, basically *any* detectable voltage).  
* I put the buzz smoothing at 250 with this setup, but it's entirely independent of the spin, so just raise it to make the couping smoother, and lower it to make it more sensitive.

I really think this is coming up on the limits of these stepped motors and the fact that that voltage off of them just isn't smooth.  It's a tradeoff between cranking smoothness and how sensitive it is to cranking backwards.

If you try v1.3.0, try lowering `SPIN_DECAY` down (you'll notice it at 10 or less for sure): it can be *just as smooth* as it was before for slow cranking.  I'm pretty confident that by just adjusting the spin_decay and buzz_smoothing it can perform as well as it did before with the diode bridge installed.

The problem is that if you make it do that, it won't respond to a change in direction, and you'll be able to reverse-crank so fast it smooths right over it.  While this is partially because of the voltage taken by the diodes, I think the deeper thing here is that we simply didn't used to care about how quickly you could reverse direction, so we could effectively smooth it much more.  The speed I want to be able to reverse-crank is about as fast as the jitter in the motor's voltage when cranking through the gears, and I don't think it can be improved much past this point.  Fine-tuning just lets you decide more closely how much smoothness you want to personally give up for fast-reversing.

Alternately, you can raise `SPIN_DECAY` to give up some slow cranking in exchange for a very fast crank reversal.

Bottom line though: I think we're bumping up on the nature of the geared motor itself, and more sensitive diodes and precise resistor selection isn't going to get any more out of it.  .  These Amazon 1n34a diodes were my favorite:

https://www.amazon.com/20-Pieces-1N34A-Germanium-Detection/dp/B079KJVKTT

Supposedly the 1n5817, per my multimeter, had a lower forward voltage (half as much!), so I might make one last experiment of leaving everything else alone and trying them one more time.

**Comment by Gubbledenut** [comment]
Thanks trying 130 now , will adjust some settings…one other thing on a real gurdy the buzz cannot work in reverse.
Reverse is only every used briefly for a staccato effect and if you look at tobie miller playing baroque gurdy she rarely uses buzz at all.

**Reply by bazmonk** [reply]
I've got no way to make the buzz not work in reverse (the code literally can't differentiate crank direction), but I understand what you mean about it not being possible on a real one.

The buzz, to be clear, is a separate mechanism to the spin.  It works off the voltage, so it benefits from the smoothing changes, but it doesn't get affected by any of the spin settings.  It's just a matter of choosing a `BUZZ_COUNTDOWN` that makes it as smooth as you want vs. sensitive to fast coups.  

I forgot to mention in all this, but the end-result of the motor circuit I'm using is that I get a similar high-end voltage as before.  The knob works more or less the same for me as it did before.

**Comment by bazmonk** [comment]
Oh, they're better.  1n5817 is where it's at!

**Reply by bazmonk** [reply]
WAY BETTER

**Reply by Gubbledenut** [reply]
Currently using in4007 looks like the 5817 is a replacement for the lower voltage in4001…I only have two of them, but have ordered some in60p which are very similar lower voltage. 
so it may just be utilising a lower voltage range on the diodes themselves which in turn will give more sensitivity? Having a similar thing with using pressure sensors with too high a limit, greatly reducing the useable range of my woodwind controllers…
I will stick with these 4007 until my in60p get here hopefully in a couple of days.

**Reply by bazmonk** [reply]
1n5817 has a forward voltage around 0.1v at the very lower end (https://www.vishay.com/docs/88525/1n5817.pdf), and I think now that's the key to the sensitivity, yeah.

It feels just like it did without the rectifier, I swear.  every little "step" of the crank registers, so it's as responsive as it can be (just smoothing motor defects now.  1.3.1 is what I'm using with it.  Note it's even more sensitive than before.

Measuring with a stopwatch, I can go as slow as about 3.75 turns per 10 seconds.  With the 1n34a just now, when I did 1.3.0, it was more like 5+ turns per 10s, and with worse response reversing crank.

I'm also getting slightly more voltage and had to adjust the knob "scaling factor" down to 2.2 from 3 to give me a range from 0 to close to 475 (the actual voltage values), which lets me turn it up enough to basically not buzz no matter how fast I go.  But that's clear evidence that more voltage is getting out now with these diodes.  

I'm very, very happy with it now.  110% happy.  I can move in rhythm with Andrey in every song I try :-).

**Comment by Gubbledenut** [comment]
ok briefly tried 130 and 131 using my 4007 bridge, I haven't messed around with the settings much, but for now I'm going back to 129 as it works nicely,,,,,when I get the other in60p diodes I will try again, and also test more without the bridge.
The jitter on 131 was very noticeable to the degree it was hard to get going smoothly...I see where you're going with this, and I think we will moving forward need to be more specific on the bridge diodes used....ideally I think if it's possible to get it working well using something generic and easily available like a 4001 ( I will get more next time I'm out to try) it will be more practical to update if anyone wants to? John does fit a diode across the motor terminal to protect the teensy from potential reverse voltage surges. Try also for fun to increase the restore across the motor terminals to increase the resistance of the crank..ideally you don't want to too easy to spin, again some real ones have a dampening screw to increase bearing resistance

**Comment by bazmonk** [comment]
I really don’t think the 1n400[1-7] family is going to work here.  

https://www.vishay.com/docs/88503/1n4001.pdf

Compare the forward voltage graph (0.6v for the 1n4001/7) with the link in my other comment for the 1n5817 (0.1v).  The unusable jitter in the 1n4007 bridge you’ve got now (and I think an 1n4001 as well) isn’t something that can be overcome.  Most basic silicon diodes are going to be in the 0.6-0.7 range.  

I noticed big improvement when I switched to the 1n5817 from 1n34as… those were 0.2v (Germanium diodes are 0.2-0.3) to give you an idea.  If that last tenth of a volt was noticeable, I don’t think you’re going to be able to work with missing another half volt, you know?

The 1n5817 should be pretty easily-available.  I know it’s all over Amazon, mouser, digikey.  Several brands make it.  It’s an in-production, common-ish diode… I even see them in most of the basic “diode assortment kits”.  There’s probably other Schottky diodes (these are the ones that seem to come in the ultra-low voltages, even lower than Germanium diodes) that would do that job, but back to my very first suspicion when John gave us this idea, the forward voltage is the most important thing to the end product.
